### PR TITLE
[BD-46] docs: update color picker readme

### PR DESCRIPTION
**Pull request by khudym**
## Description

[Issue](https://github.com/openedx/paragon/issues/2524)

### Deploy Preview

[Preview](https://deploy-preview-2646--paragon-openedx.netlify.app/components/colorpicker/)
Include a direct link to your changes in this PR's deploy preview here (e.g., a specific component page).

## Merge Checklist

* [ ] If your update includes visual changes, have they been reviewed by a designer? Send them a link to the Netlify deploy preview, if applicable.
* [ ] Does your change adhere to the documented [style conventions](https://github.com/openedx/paragon/blob/master/docs/decisions/0012-css-styling-conventions)?
* [ ] Do any prop types have missing descriptions in the Props API tables in the documentation site (check deploy preview)?
* [ ] Were your changes tested using all available themes (see theme switcher in the header of the deploy preview, under the "Settings" icon)?
* [ ] Were your changes tested in the `example` app?
* [ ] Is there adequate test coverage for your changes?
* [ ] Consider whether this change needs to reviewed/QA'ed for accessibility (a11y). If so, please add `wittjeff` and `adamstankiewicz` as reviewers on this PR.

## Post-merge Checklist

* [ ] Verify your changes were released to [NPM](https://www.npmjs.com/package/@edx/paragon) at the expected version.
* [ ] If you'd like, [share](https://github.com/openedx/paragon/discussions/new?category=show-and-tell) your contribution in [#show-and-tell](https://github.com/openedx/paragon/discussions/categories/show-and-tell).
* [ ] 🎉 🙌 Celebrate! Thanks for your contribution.

### CODE REVIEW

**Comment by brian-smith-tcril** [comment]
I was curious about this so I did a little bit of digging. It seems the issue is that the [`react-colorful`](https://github.com/omgovich/react-colorful) hex picker doesn't support invalid hex strings at all. After cloning the `react-colorful` repo and running `npm run start-demo` locally I looked at their example and found they get around this by simply not allowing invalid hex in the textbox at all

badcolor

another thing I tried was changing
https://github.com/openedx/paragon/blob/77ebd94d091e3ae45e17cb31150214a3cf553f4a/src/ColorPicker/index.jsx#L68

to be

```jsx
<HexColorPicker color={hexValid ? color : '#fff'} onChange={setColor} />
```

which fixes the `#NaNNaNNaN` problem, but still feels less than ideal

nonan

I looked into the event handling for `onChange`, and realized it wasn't being fired when moving the hue slider (if the currently selected color doesn't have a hue, so anything on the far left of the picker box). I then found that this is a known issue that has been reported on their repo https://github.com/omgovich/react-colorful/issues/206

I then tried another workaround (which feels a little hacky), instead using
```jsx
<HexColorPicker color={hexValid ? color : '#fefefe'} onChange={setColor} />
```

fefefe

and that seems to give us expected behavior at least.

I'm not sure what the best path forward for this is. But at least it seems we have a few options to consider now!

**Comment by cintnguyen** [comment]
@brian-smith-tcril
```jsx
<HexColorPicker color={hexValid ? color : '#fefefe'} onChange={setColor} />
```
I think this is a good workaround for now to ensure there's always a hue to default to. Any idea why #fefefe turns into #fff?

**Comment by brian-smith-tcril** [comment]
> Any idea why #fefefe turns into #fff?

Good question!

### tl:dr - rounding inaccuracy

## full explanation below:

The hex value entered is parsed using [`hexToHsva`](https://github.com/omgovich/react-colorful/blob/c7e87161d71be9156b0c149d0e110fa24fd7037d/src/utils/convert.ts#L14C51-L14C51).
```ts
export const hexToHsva = (hex: string): HsvaColor => rgbaToHsva(hexToRgba(hex));
```
which is calling [`hexToRgba`](https://github.com/omgovich/react-colorful/blob/c7e87161d71be9156b0c149d0e110fa24fd7037d/src/utils/convert.ts#L16-L34)
```ts
export const hexToRgba = (hex: string): RgbaColor => {
  if (hex[0] === "#") hex = hex.substring(1);

  if (hex.length < 6) {
    return {
      r: parseInt(hex[0] + hex[0], 16),
      g: parseInt(hex[1] + hex[1], 16),
      b: parseInt(hex[2] + hex[2], 16),
      a: hex.length === 4 ? round(parseInt(hex[3] + hex[3], 16) / 255, 2) : 1,
    };
  }

  return {
    r: parseInt(hex.substring(0, 2), 16),
    g: parseInt(hex.substring(2, 4), 16),
    b: parseInt(hex.substring(4, 6), 16),
    a: hex.length === 8 ? round(parseInt(hex.substring(6, 8), 16) / 255, 2) : 1,
  };
};
```
with `#fefefe`, `hexToRgba` is returning
```
{
  r: 254,
  g: 254,
  b: 254,
  a: 1,
}
```
which then goes into [`rgbaToHsva`](https://github.com/omgovich/react-colorful/blob/c7e87161d71be9156b0c149d0e110fa24fd7037d/src/utils/convert.ts#L171-L190)
```ts
export const rgbaToHsva = ({ r, g, b, a }: RgbaColor): HsvaColor => {
  const max = Math.max(r, g, b);
  const delta = max - Math.min(r, g, b);

  // prettier-ignore
  const hh = delta
    ? max === r
      ? (g - b) / delta
      : max === g
        ? 2 + (b - r) / delta
        : 4 + (r - g) / delta
    : 0;

  return {
    h: round(60 * (hh < 0 ? hh + 6 : hh)),
    s: round(max ? (delta / max) * 100 : 0),
    v: round((max / 255) * 100),
    a,
  };
};
```
which returns
```
{
  h: 0,
  s: 0,
  v: 100,
  a: 1,
}
```
which is then stored as the internal color value for the picker.

when we move the slider, `react-colorful` uses the color manipulation [`handleChange`](https://github.com/omgovich/react-colorful/blob/c7e87161d71be9156b0c149d0e110fa24fd7037d/src/hooks/useColorManipulation.ts#L47-L49)
```ts
  const handleChange = useCallback((params: Partial<HsvaColor>) => {
    updateHsva((current) => Object.assign({}, current, params));
  }, []);
```
which only modifies the `h` in the hsva object. the new internal value is then converted to hex [here](https://github.com/omgovich/react-colorful/blob/c7e87161d71be9156b0c149d0e110fa24fd7037d/src/components/HexColorPicker.tsx#L11)
```tsx
fromHsva: ({ h, s, v }) => hsvaToHex({ h, s, v, a: 1 }),
```
which uses [`hsvaToHex` from `convert.ts`](https://github.com/omgovich/react-colorful/blob/c7e87161d71be9156b0c149d0e110fa24fd7037d/src/utils/convert.ts#L67)
```ts
export const hsvaToHex = (hsva: HsvaColor): string => rgbaToHex(hsvaToRgba(hsva));
```
which calls [`hsvaToRgba`](https://github.com/omgovich/react-colorful/blob/c7e87161d71be9156b0c149d0e110fa24fd7037d/src/utils/convert.ts#L100-L117)
```ts
export const hsvaToRgba = ({ h, s, v, a }: HsvaColor): RgbaColor => {
  h = (h / 360) * 6;
  s = s / 100;
  v = v / 100;

  const hh = Math.floor(h),
    b = v * (1 - s),
    c = v * (1 - (h - hh) * s),
    d = v * (1 - (1 - h + hh) * s),
    module = hh % 6;

  return {
    r: round([v, c, b, b, d, v][module] * 255),
    g: round([d, v, v, c, b, b][module] * 255),
    b: round([b, b, d, v, v, c][module] * 255),
    a: round(a, 2),
  };
};
```
which, for `{h: ANYTHING, s: 0, v: 100, a: 1}`, returns
```
{
  r: 255,
  g: 255,
  b: 255,
  a: 1
}
```
which, if we don't pay attention to a, is `#ffffff`

**Comment by openedx-webhooks** [comment]
@khudym Even though your pull request wasn’t merged, please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/paragon&entry.867055334=https://github.com/openedx/paragon/pull/2646&entry.1484655318=https://github.com/khudym&entry.752974735=2023-09-22+13:18&entry.1917517419=2023-10-26+09:08&entry.2133058324=No) so we can improve your experience in the future.
